File: CS/WindowsApplication1/InplaceEditorHelper/ColumnInplaceEditorHelper.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Drawing;
using DevExpress.Utils;

namespace WindowsApplication1
{
    public class ColumnInplaceEditorHelper
    {
        private RepositoryItem _Item;
        private GridColumn _Column;
        GridView view;
        public ColumnInplaceEditorHelper(GridColumn column, RepositoryItem inplaceEditor)
        {
            _Column = column;
            _Item = inplaceEditor;
            view = column.View as GridView;
            view.CustomDrawColumnHeader += view_CustomDrawColumnHeader;
            view.MouseDown += view_MouseDown;
            view.Layout += new EventHandler(view_Layout);
        }

        void view_Layout(object sender, EventArgs e)
        {
            CloseEditor();
        }

        private object _EditValue;
        public object EditValue
        {
            get { return _EditValue; }
            set
            {
                _EditValue = value;
            }
        }

        private BaseEdit _ActiveEditor;
        public BaseEdit ActiveEditor
        {
            get { return _ActiveEditor; }
            set { _ActiveEditor = value; }
        }

        void view_CustomDrawColumnHeader(object sender, ColumnHeaderCustomDrawEventArgs e)
        {
            if (e.Column == _Column)
            {
                e.Info.Caption = string.Empty;
                e.Painter.DrawObject(e.Info);
                e.Handled = true;
                DrawEditorHelper.DrawColumnInplaceEditor(e, _Item, EditValue, GetRightIndent());
            }
        }

        public int GetRightIndent()
        {
            return _Column.OptionsColumn.AllowSort != DevExpress.Utils.DefaultBoolean.False || _Column.OptionsFilter.AllowFilter ? 25 : 0;
        }
        void view_MouseDown(object sender, MouseEventArgs e)
        {
            CloseEditor();
            Rectangle editorBounds;
            if (ClickInEditor(e, out editorBounds)) 
            {
                ShowEditor(editorBounds);
                DXMouseEventArgs.GetMouseArgs(e).Handled = true;
            }
        }
        private bool ClickInEditor(MouseEventArgs e, out Rectangle editorBounds)
        {
            editorBounds = Rectangle.Empty;
            GridViewInfo vi = view.GetViewInfo() as GridViewInfo;
            GridColumnInfoArgs columnInfo = CalcColumnHitInfo(e.Location, vi.ColumnsInfo);

            if (columnInfo != null && columnInfo.Column == _Column)
            {
                editorBounds = DrawEditorHelper.GetEditorBounds(columnInfo.Bounds, GetRightIndent());
                return editorBounds.Contains(e.Location);
            }
            return false;
        }

        protected virtual GridColumnInfoArgs CalcColumnHitInfo(Point pt, GridColumnsInfo cols)
        {
            foreach (GridColumnInfoArgs ci in cols)
            {
                if (!ci.Bounds.IsEmpty && IntInRange(pt.X, ci.Bounds.Left, ci.Bounds.Right))
                {
                    if (ci.Type == GridColumnInfoType.EmptyColumn) continue;
                    return ci;
                }
            }
            return null;
        }

        protected bool IntInRange(int x, int left, int right)
        {
            if (right < left)
            {
                int temp = left;
                left = right;
                right = temp;
            }
            return (x >= left && x < right);
        }

        private void ShowEditor(Rectangle bounds)
        {
            ActiveEditor = _Item.CreateEditor();
            ActiveEditor.Properties.LockEvents();
            ActiveEditor.Parent = view.GridControl;
            ActiveEditor.Properties.Assign(_Item);
            ActiveEditor.Properties.AutoHeight = false;
            ActiveEditor.Location = bounds.Location;
            ActiveEditor.Size = bounds.Size;
            ActiveEditor.CreateControl();
            ActiveEditor.EditValue = EditValue;
            ActiveEditor.SendMouse(ActiveEditor.PointToClient(Control.MousePosition), Control.MouseButtons);
            ActiveEditor.Properties.UnLockEvents();
        }

        private void CloseEditor()
        {
            if (ActiveEditor != null)
            {
                EditValue = ActiveEditor.EditValue;
                ActiveEditor.Dispose();
                ActiveEditor = null;
            }
        }
        void editor_Leave(object sender, EventArgs e)
        {
            CloseEditor();
        }
    }
}

File: VB/WindowsApplication1/InplaceEditorHelper/ColumnInplaceEditorHelper.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Drawing
Imports DevExpress.Utils
Imports System.Runtime.InteropServices

Namespace WindowsApplication1

    Public Class ColumnInplaceEditorHelper

        Private _Item As RepositoryItem

        Private _Column As GridColumn

        Private view As GridView

        Public Sub New(ByVal column As GridColumn, ByVal inplaceEditor As RepositoryItem)
            _Column = column
            _Item = inplaceEditor
            view = TryCast(column.View, GridView)
            AddHandler view.CustomDrawColumnHeader, AddressOf view_CustomDrawColumnHeader
            AddHandler view.MouseDown, AddressOf view_MouseDown
            AddHandler view.Layout, New EventHandler(AddressOf Me.view_Layout)
        End Sub

        Private Sub view_Layout(ByVal sender As Object, ByVal e As EventArgs)
            CloseEditor()
        End Sub

        Private _EditValue As Object

        Public Property EditValue As Object
            Get
                Return _EditValue
            End Get

            Set(ByVal value As Object)
                _EditValue = value
            End Set
        End Property

        Private _ActiveEditor As BaseEdit

        Public Property ActiveEditor As BaseEdit
            Get
                Return _ActiveEditor
            End Get

            Set(ByVal value As BaseEdit)
                _ActiveEditor = value
            End Set
        End Property

        Private Sub view_CustomDrawColumnHeader(ByVal sender As Object, ByVal e As ColumnHeaderCustomDrawEventArgs)
            If e.Column Is _Column Then
                e.Info.Caption = String.Empty
                e.Painter.DrawObject(e.Info)
                e.Handled = True
                DrawColumnInplaceEditor(e, _Item, EditValue, GetRightIndent())
            End If
        End Sub

        Public Function GetRightIndent() As Integer
            Return If(_Column.OptionsColumn.AllowSort <> DefaultBoolean.False OrElse _Column.OptionsFilter.AllowFilter, 25, 0)
        End Function

        Private Sub view_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs)
            CloseEditor()
            Dim editorBounds As Rectangle
            If ClickInEditor(e, editorBounds) Then
                ShowEditor(editorBounds)
                DXMouseEventArgs.GetMouseArgs(e).Handled = True
            End If
        End Sub

        Private Function ClickInEditor(ByVal e As MouseEventArgs, <Out> ByRef editorBounds As Rectangle) As Boolean
            editorBounds = Rectangle.Empty
            Dim vi As GridViewInfo = TryCast(view.GetViewInfo(), GridViewInfo)
            Dim columnInfo As GridColumnInfoArgs = CalcColumnHitInfo(e.Location, vi.ColumnsInfo)
            If columnInfo IsNot Nothing AndAlso columnInfo.Column Is _Column Then
                editorBounds = GetEditorBounds(columnInfo.Bounds, GetRightIndent())
                Return editorBounds.Contains(e.Location)
            End If

            Return False
        End Function

        Protected Overridable Function CalcColumnHitInfo(ByVal pt As Point, ByVal cols As GridColumnsInfo) As GridColumnInfoArgs
            For Each ci As GridColumnInfoArgs In cols
                If Not ci.Bounds.IsEmpty AndAlso IntInRange(pt.X, ci.Bounds.Left, ci.Bounds.Right) Then
                    If ci.Type = GridColumnInfoType.EmptyColumn Then Continue For
                    Return ci
                End If
            Next

            Return Nothing
        End Function

        Protected Function IntInRange(ByVal x As Integer, ByVal left As Integer, ByVal right As Integer) As Boolean
            If right < left Then
                Dim temp As Integer = left
                left = right
                right = temp
            End If

            Return x >= left AndAlso x < right
        End Function

        Private Sub ShowEditor(ByVal bounds As Rectangle)
            ActiveEditor = _Item.CreateEditor()
            ActiveEditor.Properties.LockEvents()
            ActiveEditor.Parent = view.GridControl
            ActiveEditor.Properties.Assign(_Item)
            ActiveEditor.Properties.AutoHeight = False
            ActiveEditor.Location = bounds.Location
            ActiveEditor.Size = bounds.Size
            ActiveEditor.CreateControl()
            ActiveEditor.EditValue = EditValue
            ActiveEditor.SendMouse(ActiveEditor.PointToClient(Control.MousePosition), Control.MouseButtons)
            ActiveEditor.Properties.UnLockEvents()
        End Sub

        Private Sub CloseEditor()
            If ActiveEditor IsNot Nothing Then
                EditValue = ActiveEditor.EditValue
                ActiveEditor.Dispose()
                ActiveEditor = Nothing
            End If
        End Sub

        Private Sub editor_Leave(ByVal sender As Object, ByVal e As EventArgs)
            CloseEditor()
        End Sub
    End Class
End Namespace
